### PR TITLE
CBG-803: Support for OIDC provider config refresh

### PR DESCRIPTION
**Pull request by sarathkumarsivan**
Changes to support OpenID Connect Provider Metadata

### CODE REVIEW

File: auth/oidc.go [hunk 1]
```diff
@@ -71,6 +75,20 @@ type OIDCClient struct {
 
 	// Verifier provides verification for ID Tokens.
 	Verifier *oidc.IDTokenVerifier
+
+	// A mutual exclusion lock for provider metadata.
+	mux sync.Mutex
```
Comment (bbrks) [inline auth/oidc.go]: Typically name locks after the resource they affect, i.e.:

```suggestion
	providerLock sync.Mutex
```

File: auth/oidc.go [hunk 2]
```diff
@@ -423,3 +456,50 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
...
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				base.Tracef(base.KeyAuth, "Synchronizing provider config for issuer %s ...", base.UD(issuer))
+				provider, err := oidc.NewProvider(context.Background(), issuer)
+				if err != nil {
+					base.Errorf("Failed synchronizing provider config for issuer %s, Error: %v", base.UD(issuer), err.Error())
+				} else {
+					provider, err = oidc.NewProvider(context.Background(), issuer)
```
Comment (bbrks) [inline auth/oidc.go]: Haven't we already done this work on line 489?

File: auth/oidc.go [hunk 3]
```diff
@@ -46,7 +46,11 @@ const (
 	discoveryConfigPath = "/.well-known/openid-configuration"
 )
 
-var OIDCDiscoveryRetryWait = 500 * time.Millisecond
+var (
+	OIDCDiscoveryRetryWait = 500 * time.Millisecond
+
+	providerConfigSyncInterval = time.NewTicker(1 * time.Second)
```
Comment (bbrks) [inline auth/oidc.go]: Refreshing each provider once per second seems SUPER aggressive...

For comparison, the old library had a lower bound of 1 minute, and an upper bound of 24 hours, and used them like this:

https://github.com/coreos/go-oidc/blob/e860bd55bfa7d7cb35d30d26a167982584f616b0/oidc/provider.go#L591-L604

File: auth/oidc.go [hunk 4]
```diff
@@ -254,7 +281,7 @@ func (op *OIDCProvider) DiscoverConfig() (verifier *oidc.IDTokenVerifier, endpoi
 		base.Infof(base.KeyAuth, "Fetching provider config from explicitly defined discovery endpoint: %s", base.UD(op.DiscoveryURI))
 		metadata, err := op.FetchCustomProviderConfig(discoveryURL)
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, shouldSync, err
```
Comment (bbrks) [inline auth/oidc.go]:
```suggestion
			return nil, nil, false, err
```

File: auth/oidc.go [hunk 5]
```diff
@@ -423,3 +456,50 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
...
+		if client.provider != nil {
+			provider = client.provider
+			client.mux.Unlock()
+			break
+		}
+		base.Infof(base.KeyAuth, "Getting provider config ready; waiting for discovery sync.")
+		client.mux.Unlock()
+		time.Sleep(100 * time.Millisecond)
+	}
+	return
```
Comment (bbrks) [inline auth/oidc.go]: Why do we need to loop whilst locking/unlocking to check? Can we not use an RWMutex for synchronization? (i.e. many readers blocked waiting for one writer to unlock)

File: auth/oidc.go [hunk 6]
```diff
@@ -423,3 +456,50 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
...
+}
+
+func (client *OIDCClient) SetProvider(provider *oidc.Provider) {
+	client.mux.Lock()
+	client.provider = provider
+	client.mux.Unlock()
+}
+
+func (client *OIDCClient) SyncProviderConfig(issuer string) {
```
Comment (bbrks) [inline auth/oidc.go]: Do we have an approach to test the provider sync/refresh mechanism?

I think it should be feasible to write a test provider that changes configuration and override the sync time for testing purposes.

**Review by bbrks — CHANGES_REQUESTED** [review]

**Comment by sarathkumarsivan** [comment]
Closing this pul request; the changes can be tracked under the new PR.

**Pull request closed without merging**